Box Blur Improvements

bmatherly opened the pull request:
This implementation has 2x performance of the boxblur filter.
It also supports slices.

Use this new box blur function in pillar_image filter
Implement a new box_blur filter using this function

Performance comparison between old and new filters:

**boxblur (old)**
```
time ./Shotcut/Shotcut.app/melt noise out=500 -filter volume:0 -filter boxblur blur=40 -consumer null terminate_on_pause=1 real_time=-1

real	0m24.335s
user	0m24.239s
sys	0m0.070s
```

**box_blur (new)**
```
time ./Shotcut/Shotcut.app/melt noise out=500 -filter volume:0 -filter box_blur -consumer null terminate_on_pause=1 real_time=-1

real	0m6.789s
user	0m12.598s
sys	0m0.139s
```

For discussion:
* Do we want to continue to add image processing functions to the mlt_image module? Or maybe they should go some where else?
* Different name? "box_blur" is not very different from "boxblur".
* Deprecate the old boxblur filter?
* The h/vradius parameters units are: "The horizontal blur radius as a percent of the image size. 100% results in a blur radius of 10% of the image width.". Do we like this, or something else?

ddennedy commented on src/framework/mlt_image.c:
```diff
@@ -425,3 +426,215 @@ void mlt_image_format_planes( mlt_image_format format, int width, int height, vo
+	int radius;
+} blur_slice_desc;
+
+static int blur_h_proc(int id, int index, int jobs, void* data)
```
Image processing functions do not belong in the framework.

bmatherly replied:
I can accept that. The idea was inspired by FFmpeg which has a gaussian function in swcale. It is part of the public interface, and multiple filters use it.
https://github.com/FFmpeg/FFmpeg/blob/069f7831a2bd1545080121fb6d0773191cd61d70/libswscale/swscale.h#L351

Maybe someday there could be a module in the framework for useful processing functions. I will make a separate proposal if I come up with something.

ddennedy replied:
This idea is a little complicated within MLT proper. There could be a libmltimgproc library sub-project added that would be a better fit. But then, is there not already enough image processing libraries? And why go into this trouble when all image processing must be completely revisited for higher than 8-bit processing?

ddennedy commented on src/modules/plus/filter_pillar_echo.c:
```diff
@@ -414,7 +251,11 @@ static int filter_get_image( mlt_frame frame, uint8_t **image, mlt_image_format
+	if (blur != 0) {
+		struct mlt_image_s img;
+		mlt_image_set_values( &img, dst, *format, *width, *height );
+		mlt_image_box_blur( &img, blur, blur );
```
To avoid code duplication the pillar_echo filter should encapsulate a blur filter that can be changed from its default via a string property (e.g. "blur_filter") including properties that begin with "blur." for example to set properties on the nested filter. This is a pattern we have used elsewhere.

bmatherly replied:
I am familiar with the service encapsulation method. I do not like it for this case because of the extra resource overhead and complication:
* Requires the construction of an entire extra service and passing parameters using the properties
* Requires constructing an extra frame to pass the image into the service
* Does not support frame threading unless the parameters are attached to the frame and interpreted by the encapsulated filter
All this extra complication and overhead - and all I want to do is have a single function that is shared by two files.

I will move this function into a shared file at the module level as we have done for other modules:
https://github.com/mltframework/mlt/tree/master/src/modules/plus/ebur128
https://github.com/mltframework/mlt/blob/master/src/modules/plusgpl/utils.h
https://github.com/mltframework/mlt/blob/master/src/modules/qt/graph.h

bmatherly commented:
I moved the blur function out of mlt_image

Also added slice processing to the scaling operation in pilar_echo.

Pillar echo improvements for this command on a 4 core machine:
`time melt noise out=500 -filter volume:0 -filter pillar_echo rect="5% 5% 90% 90%" -consumer null  terminate_on_pause=1 real_time=-1`

Before:
real	0m22.636s
user	0m22.599s

After
real	0m9.621s
user	0m23.985s

ddennedy commented on src/modules/core/filter_box_blur.yml:
```diff
@@ -0,0 +1,35 @@
+      The vertical blur radius as a percent of the image size.
+      100% results in a blur radius of 10% of the image width.
+    type: integer
+    unit: pixels
```
unit should be percentage (or percent)

bmatherly replied:
done

ddennedy commented on src/modules/core/filter_box_blur.c:
```diff
@@ -0,0 +1,92 @@
+	// Convert from percent to pixels
+	hradius = hradius * (double)profile->width * mlt_profile_scale_width( profile, *width ) / 1000.0;
+	hradius = MAX(lrint(hradius), 0);
+	vradius = vradius * (double)profile->width * mlt_profile_scale_width( profile, *width ) / 1000.0;
```
Use `mlt_profile_scale_height()` here

bmatherly replied:
I want the scale factor to be the same for H and V so the user does not have to guess-and-check the values if they want even h/v blur size. I refactored this code to make it more obvious.

ddennedy commented on src/modules/core/filter_box_blur.yml:
```diff
@@ -0,0 +1,35 @@
+    description: >
+      The horizontal blur radius as a percent of the image size.
+      100% results in a blur radius of 10% of the image width.
+    type: integer
```
I am confused a little about integer and percent here because it is being get-ed in the code as a double, and some projects such as frei0r use [0.0, 1.0] as a range for a percentage. I understand it a little better from the description but still kind of awkward. Since there is no maximum, then I can use 1000% for 100% of the image dimension. So, why not keep it simpler and use [0.0, 1.0] or [0.0, 100.0] where 100% = 100% and let me use higher precision than 0.1% of the image dimension? I think you are trying to map the most usable portion of the range into [0, 100] but also not make it limited. I think that makes sense for a UI, and this can simplify a UI for this; but not so sure it makes sense here. I will not object but at the very least, change the type to float.

bmatherly replied:
>  I think you are trying to map the most usable portion of the range into [0, 100] but also not make it limited. I think that makes sense for a UI, and this can simplify a UI for this

You understand my dilemma perfectly. If I set the maximum range to be 100% of the image width, then in the UI I need to scale that down by 1/10th because 100% image blur is not useful in the application (and all the useful values would be in the first 10% of the slider). 

I was hoping you would offer a stronger opinion and then I would change it to whatever your preference is. I agree with changing the type to float. But let me know if you think of a better range/scale for this parameter.

bmatherly replied:
Changed the type to float.

ddennedy commented on src/modules/core/filter_pillar_echo.c:
```diff
@@ -0,0 +1,314 @@
+	int linesize = src->width * 4;
+	uint8_t* d = dst->data + (slice_line_start * linesize);
+	int y = 0;
+	for ( y = slice_line_start; y < slice_line_end; y++ )
```
Just a FYI for a while now we have supported using C99 so you can declare within a for loop, for example.

bmatherly replied:
Nice! Done.

ddennedy commented on src/modules/core/filter_pillar_echo.c:
```diff
@@ -0,0 +1,314 @@
+	}
+
+	*format = mlt_image_rgba;
+	error = mlt_frame_get_image( frame, image, format, width, height, 1 );
```
Here you are requesting a writable image from upstream, but that is not required. That can cause some producers to be slightly less efficient.

bmatherly replied:
Good to know. Fixed!

bmatherly commented on src/modules/core/filter_box_blur.yml:
```diff
@@ -0,0 +1,35 @@
+type: filter
+identifier: box_blur
+title: Box Blur
+version: 3
```
Done

bmatherly commented:
I believe all comments are resolved. Let me know if you want any other changes.
